### ISSUE TITLE
bug: Invalid signature when signing with hardware wallet

### ISSUE THREAD

**Issue by wyze** [wagmi-dev/viem#537]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.3.30

### Current Behavior

Signature is invalid.

### Expected Behavior

Signature is valid.

### Steps To Reproduce

1. Connect Wallet via MetaMask
2. Sign message
3. Valid: Yes

1. Connect Hardware (I used Ledger) Wallet via Metamask
2. Sign message
3. Valid: No

### Link to Minimal Reproducible Example (StackBlitz, CodeSandbox, GitHub repo etc.)

https://github.com/wyze/viem-verify-sign-issue

### Anything else?

It works as expected when using a regular MetaMask wallet. When I try to sign a message using my Ledger via MetaMask, the signature comes back invalid.

**Comment by jxom** [comment]
Can you confirm, does this also happen using the [`verifyMessage` util](https://viem.sh/docs/utilities/verifyMessage.html) (not the public client one)?

**Comment by jxom** [comment]
Hmm ok interesting, let me do some digging

**Comment by wyze** [comment]
I did some digging as well, and seeing how ethers has handled it in the past. The recovery bit from the hardware wallet is coming in as 0, so its passed in as -27, where as from a regular MetaMask wallet the bit is `b0` (27), so it gets passed in as 0.

https://github.com/wagmi-dev/viem/blob/55ca10ff918a68954e2be84b956dd19c7f035b30/src/utils/signature/recoverPublicKey.ts#LL27C28-L27C28

On that line above I updated it like so:

```diff
  const publicKey = secp256k1.Signature.fromCompact(
    signatureHex.substring(2, 130),
  )
-    .addRecoveryBit(v - 27)
+    .addRecoveryBit(v < 27 ? v : v - 27)
    .recoverPublicKey(hashHex.substring(2))
    .toHex(false)
```

Based on the logic found here: https://github.com/ethers-io/ethers.js/blob/bf65ddbff0036f6eb8e99c145f30edff157687f5/packages/bytes/src.ts/index.ts#L343-L350

So this would make the `verifyMessage` util say `Yes` (valid), but still fails on the publicClient.verifyMessage.
